crates/media-parser/src/stream.rs

```diff
@@ -1,65 +1,537 @@
+   fn read_at_offset(&self, buf: &mut [u8], offset: u64) -> std::io::Result<usize> {
+      FileExt::seek_read(self, buf, offset)
+   }
+}
+
+// Constants
+/// HTTP status code for partial content (Range request success)
+const HTTP_PARTIAL_CONTENT: u16 = 206;
+
+/// Copies bytes from `src` into `dst` and returns the count.
+fn copy_into(dst: &mut [u8], src: &[u8]) -> usize {
+   let len = dst.len().min(src.len());
+   dst[..len].copy_from_slice(&src[..len]);
+   len
+}
+
+/// Provides a interface for reading data at arbitrary offsets
+///
+/// # Contract
+///
+/// - `read_at` reads up to `buf.len()` bytes starting at `offset`
+/// - Returns the number of bytes read (may be less than `buf.len()` if EOF)
+/// - Returns `0` if `offset >= size()`
+/// - Partial reads are allowed
 
-/// Trait for reading media streams from various sources
 #[async_trait]
 pub trait StreamReader: Send + Sync {
-   /// Read data at the specified offset into the provided buffer.
+   /// Reads data at the specified offset into the buffer.
+   ///
+   /// Reads up to `buf.len()` bytes starting at `offset`. Returns the number of bytes
+   /// actually read, which may be less than `buf.len()` if EOF is reached. Returns `0`
+   /// if `offset >= size()` or if `buf.is_empty()`.
    async fn read_at(&self, offset: u64, buf: &mut [u8]) -> Result<usize>;
 
-   /// Get the total size of the stream.
+   /// Returns the total size of the stream in bytes.
    async fn size(&self) -> Result<u64>;
 }
 
-/// Local file stream reader
-pub struct FileStreamReader {}
+/// `StreamReader` implementation backed by a local file handle.
+///
+/// # Examples
+///
+/// ```no_run
+/// use media_parser::{FileStreamReader, StreamReader};
+///
+/// # async fn example() -> media_parser::Result<()> {
+/// let reader = FileStreamReader::new("video.mp4")?;
+/// let size = reader.size().await?;
+///
+/// let mut buffer = vec![0u8; 1024];
+/// let bytes_read = reader.read_at(0, &mut buffer).await?;
+/// # Ok(())
+/// # }
+/// ```
+pub struct FileStreamReader {
+   file: Arc<std::fs::File>,
+   cached_size: OnceLock<u64>,
+}
 
 impl FileStreamReader {
-   pub fn new<P: AsRef<std::path::Path>>(_path: P) -> Self {
-      Self {}
+   /// Opens a file at the given path for random-access reading.
+   ///
+   /// Returns an error if the file does not exist or cannot be opened.
+   pub fn new<P: AsRef<Path>>(path: P) -> Result<Self> {
+      let file = Arc::new(std::fs::File::open(path).map_err(MediaParserError::Io)?);
+      Ok(Self {
+         file,
+         cached_size: OnceLock::new(),
+      })
+   }
+
+   /// Performs a blocking `read_at` directly into the provided buffer.
+   fn sync_read_into(file: &std::fs::File, offset: u64, buf: &mut [u8]) -> Result<usize> {
+      let mut read_total = 0usize;
+      while read_total < buf.len() {
+         let n = file
+            .read_at_offset(&mut buf[read_total..], offset + read_total as u64)
+            .map_err(MediaParserError::Io)?;
+         if n == 0 {
```

Comment:
Potential infinite loop: If `read_at_offset` keeps returning 0.
Should we add a retry limit or detect when no progress is being made?